### PR TITLE
fix issue 18784 - cpp constructor returns void on Posix

### PR DESCRIPTION
**Pull request by SSoulaimane**

### CODE REVIEW

File: src/dmd/e2ir.d [hunk 1]
```diff
@@ -514,7 +514,19 @@ if (!irs.params.is64bit) assert(tysize(TYnptr) == 4);
...
     }
 
-    if (retmethod == RET.stack)
+    const isCPPCtor = fd && fd.linkage == LINK.cpp && fd.isCtorDeclaration();
+    if (isCPPCtor &&
+        (irs.params.isLinux ||
+         irs.params.isOSX ||
+         irs.params.isFreeBSD ||
+         irs.params.isDragonFlyBSD ||
+         irs.params.isSolaris))
```
Comment (Geod24) [inline src/dmd/e2ir.d]: Perhaps some `isPOSIX` function would be better suited here. The C++ side has [this](https://github.com/dlang/dmd/blob/9590dc86eca66e55cf9e28ca600e69589ff2ac35/src/dmd/root/object.h#L12).

File: src/dmd/e2ir.d [hunk 2]
```diff
@@ -514,7 +514,19 @@ if (!irs.params.is64bit) assert(tysize(TYnptr) == 4);
...
+    {
+        // CPP constructor returns void on Posix
+        // https://itanium-cxx-abi.github.io/cxx-abi/abi.html#return-value-ctor
+        e = el_combine(e, el_copytree(ethis));
```
Comment (WalterBright) [inline src/dmd/e2ir.d]: Merely copying the tree causes problems if the tree has side effects.

Comment (SSoulaimane) [reply src/dmd/e2ir.d]: Changed.

**Comment by Geod24** [comment]
Please also add a test where the `ctor` has a side effect

**Comment by SSoulaimane** [comment]
> Please also add a test where the `ctor` has a side effect

Why? It's not related to this.

**Review by thewilsonator — APPROVED** [review]

**Pull request merged**